livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -206,20 +231,37 @@ private SessionInfo getSessionInfo(int sessionId) throws LivyException {
     return SessionInfo.fromJson(callRestAPI("/sessions/" + sessionId, "GET"));
   }
 
-  public InterpreterResult interpret(String code, boolean displayAppInfo)
+  private LivyVersion getLivyVersion() throws LivyException {
+    return new LivyVersion((LivyVersionResponse.fromJson(callRestAPI("/version", "GET")).version));
+  }
+
+  public InterpreterResult interpret(String code, String paragraphId, boolean displayAppInfo)
       throws LivyException {
-    StatementInfo stmtInfo = executeStatement(new ExecuteRequest(code));
-    // pull the statement status
-    while (!stmtInfo.isAvailable()) {
-      try {
-        Thread.sleep(1000);
-      } catch (InterruptedException e) {
-        LOGGER.error("InterruptedException when pulling statement status.", e);
-        throw new LivyException(e);
+    try {
+      StatementInfo stmtInfo = executeStatement(new ExecuteRequest(code));
+      if (paragraphId != null) {
+        paragraphId2StmtIdMapping.put(paragraphId, stmtInfo.id);
+      }
+      // pull the statement status
+      while (!stmtInfo.isAvailable()) {
+        try {
+          Thread.sleep(1000);
```

Comment:
livy don't have push mode to notify the statement is finished. So here I have to block here and pull status periodically, I add one new configuration `zeppelin.livy.pull_status.interval.millis` for it.